### PR TITLE
feat: allow datacenter ips

### PR DESCRIPTION
**Pull request by nikkaroraa**
# Task:
Exclude datacenter IPs from VPNs

## Description
Fixes ENG-1944

## Type of change

- [x] New feature
- [ ] Bug fix
- [ ] Testing code
- [ ] Document update or config files

### CODE REVIEW

File: packages/frontend/middleware.ts [hunk 1]
```diff
@@ -19,17 +21,16 @@ export async function middleware(request: NextRequest) {
 
   if (ip && !isIPWhitelisted) {
     const redisData = await redis.get(ip)
-    const isIPBlocked = !!redisData
+    const isIPBlocked = redisData === BLOCKED_IP_VALUE
```
Comment (nikkaroraa) [inline packages/frontend/middleware.ts:24]: older values don't make any sense now, similar to refreshing the older blocked ips list

Comment (aleone) [reply packages/frontend/middleware.ts:24]: Yeah thats fair.  

The only reason for the storage of these was to reduce risk of hitting api limits on ipqs.  

In future may want to have a reset of values at some cadence and store the timestamp when it is saved to handle changes in ip ranges/data centers/vpns etc. Something like this below terrible pseudocode:

``` isIPBlocked redisData[0] === BLOCKED_IP_VALUE && (Date.now() - redisData[1]) / 86400000< 30 ```

then later just setting 2 values, maybe array or something, not super familiar on what possible there, to redis:

``` await redis.set(ip, [BLOCKED_IP_VALUE, Date.now()]) ```

File: packages/frontend/src/constants/index.ts [hunk 2]
```diff
@@ -70,6 +70,9 @@ export const BLOCKED_COUNTRIES = [
 ]
 export const ALLOW_WITHDRAWALS = ['CA']
 
+export const WHITELISTED_DATACENTER_ORGANIZATIONS = ['amazon', 'google', 'azure', 'oracle', 'softlayer', 'cloudflare']
```
Comment (aleone) [inline packages/frontend/src/constants/index.ts:73]: should this be ENV for easier adjustment, etc?

**Review by aleone — APPROVED** [review]
generally lgtm

1 comment on future re resetting redis db + storing a time stamp

**Review by alpinechicken — COMMENTED** [review]
lgtm - might be good to do a test to check if gcp or aws can get through

File: packages/frontend/middleware.ts (hunk 1 above)
Comment (nikkaroraa) [reply packages/frontend/middleware.ts:24]: yeah agreed thats better. i'll make another PR to implement this.

File: packages/frontend/src/constants/index.ts (hunk 2 above)
Comment (nikkaroraa) [reply packages/frontend/src/constants/index.ts:73]: agreed

File: packages/frontend/src/server/ipqs.ts [hunk 3]
```diff
@@ -62,17 +64,27 @@ async function get_IPQ_URL(url: string): Promise<ApiResponse | null> {
...
-    return true
-  } else {
-    return false
+  const isSuspectedToBeVPN = ipResult.vpn === true || ipResult.tor === true
+  if (isSuspectedToBeVPN) {
+    const org = ipResult.organization
+
+    // exclude datacenter IPs
+    // organization should include one of WHITELISTED_DATACENTER_ORGANIZATIONS
+    const isWhitelistedOrg = WHITELISTED_DATACENTER_ORGANIZATIONS.some((whitelistedOrg) =>
```
[the remark this answers is not in the document]
Comment (nikkaroraa) [reply packages/frontend/src/server/ipqs.ts:79]: you can hardcode the ipAddress to be one from [this sheet](https://docs.google.com/spreadsheets/u/1/d/1vzXb1u4cQlpzKCcyyl44HAac8ay6to2wpDShgZVF-hQ/edit#gid=0)

File: packages/frontend/src/constants/index.ts (hunk 2 above)
Comment (nikkaroraa) [reply packages/frontend/src/constants/index.ts]: done: https://github.com/opynfinance/squeeth-monorepo/pull/1012/commits/e075f27bd58e58de5711df971f55114375abfa1a

File: packages/frontend/middleware.ts (hunk 1 above)
Comment (nikkaroraa) [reply packages/frontend/middleware.ts]: implemented it here: https://github.com/opynfinance/squeeth-monorepo/pull/1015

**Review by alpinechicken — APPROVED** [review]
changes lgtm

**Pull request merged**